### PR TITLE
use webpack-plugin-serve

### LINKED ISSUE #6: Use webpack-plugin-serve instead of webpack-dev-server
Please consider trying https://github.com/shellscape/webpack-plugin-serve as an alternative to webpack-dev-server. It's leaps and bounds better in quite a few areas, much lighter, doesn't require learning a separate CLI or API, more flexible for users, and will be a lot quicker for folks especially with static sites. We've got a fairly nice feature comparison grid/notes page here: https://github.com/shellscape/webpack-plugin-serve/blob/master/.github/FEATURES.md

### PR DESCRIPTION
**Pull request by shellscape**
This PR resolves #6, proposes to use [`webpack-plugin-serve`](https://github.com/shellscape/webpack-plugin-serve), a smaller, faster, more modern, and more flexible dev-server for webpack. Using this plugin also removes the need for a separate CLI and gives users error/build status overlays and progress indicators out of the box. Please see the associated issues for a link to the Feature Comparison Grid and additional information.

In addition to removing the need for webpack-dev-server, this PR removes the need for webpack-cli by leveraging [`webpack-nano`](https://www.npmjs.com/package/webpack-nano) - which is 90% smaller than webpack-cli, and because it negates the need for useless processing typically results in faster builds.

Overall this is an experience upgrade for your users, and a faster, lighter node_modules for the boilerplate. I do hope you'll consider merging this PR 🍻

### CODE REVIEW

**Comment by tr1s** [comment]
Hey @shellscape, thank you for your contributions! Few questions.

Is there anything to worry about when it comes to the types of projects that are being built off this boilerplate? For example, building a static website vs building a web app? Are there reasons why `webpack-dev-server` has not implemented things you've mentioned here? Let me rephrase a bit, what are some use cases in which using `webpack-dev-server` / `webpack-nano` would be better than using `webpack-plugin-serve`? Also I noticed you took the `--open` flag out, will it still open the project in browser upon `npm start`?

My worry is that I stray too far from the Webpack documentation and confuse users of this boilerplate. But I do like the idea of using tools more modern and fast. But also love things to be robust and stable. Help me understand a little bit more!

I will test this out shortly, but also, if I do implement this, can you please go through this PR and update the documentation on Webpack Development Server? A big part of this boilerplate is to provide clear explanations to readers on how and why everything works along with the working code.

Cheers 🙂

**Comment by shellscape** [comment]
No worries! Thanks for being open to the change!

> Is there anything to worry about when it comes to the types of projects that are being built off this boilerplate?

Not at all. We have recipes that cover Vue, Angular, and React, and we're working with a contributor right now on a React SSR recipe.

> Are there reasons why webpack-dev-server has not implemented things you've mentioned here?

That's a very loaded question :) As a former webpack core member I have some very strong opinions as to why that may be, and they aren't positive nor flattering to the maintainers, so it's better that I withhold comments on that. I do feel we've been able to innovate faster and better with webpack-plugin-serve, and provide feature parity with better support and flexibility.

> Let me rephrase a bit, what are some use cases in which using webpack-plugin-serve / webpack-nano would be better than using webpack-dev-server? 

I think the feature comparison and info at the bottom of this page https://github.com/shellscape/webpack-plugin-serve/blob/master/.github/FEATURES.md really covers it. Users get a better experience overall: faster HMR response in the browser, curated overlays for errors, build status, and build progress, a more flexible server (if they want to tinker with it), better options, and far fewer problems with HMR. We've basically built this plugin to benefit from all of the mistakes that webpack-dev-server has made in the past, and in some regards, continues to make. On my own machine using your configs as a baseline, I save several seconds starting up the development server with webpack-plugin-serve versus webpack-dev-server. Using webpack-nano over webpack-cli saves me some additional time for the production build as well, but the most important savings comes for development time.

> Also I noticed you took the --open flag out, will it still open the project in browser upon npm start?

That's still supported, just in a different place. Please see this line in the changed source https://github.com/tr1s/tris-webpack-boilerplate/pull/9/files#diff-f829eca9ad810a385b046aa894f64f1eR13.

> My worry is that I stray too far from the Webpack documentation

While the webpack documentation is useful, it tends to drown out a lot of alternatives that are extremely useful, and often far better for the user/developer experience. webpack-merge is a great example of a tool for webpack that is not documented on their official site (merely mentioned), but is one of the most popular, and your configs make use of it. Webpack users would really benefit from a lot of the non-blessed alternatives out there that aren't part of the official Webpack organization. 

> can you please go through this PR and update the documentation on Webpack Development Server?

Absolutely!


Please feel free to ask more questions, I'm happy to continue the discussion!

**Comment by tr1s** [comment]
@shellscape Thanks so much for the answers. Let's give it a go! Please update the docs and I'll merge this PR. I'll ping you if myself or anyone else run into issues regarding these updates specifically. Also, I'll do some more reading of your docs tonight.

Side note: took it for a spin, what's up with the [flash](https://i.imgur.com/eoPN4pN.jpg) upon reloading? Reminds me of the classic film cue mark. Is this necessary? I feel like it's a bit jarring.

I may have some more questions eventually after I read into it more. But for now, it works and is faster so that's great. Thanks for being open on keeping this discussion going.

**Comment by shellscape** [comment]
> what's up with the flash upon reloading?

That's the build progress indicator/overlay. Depending on workstation (multiple screens, etc) and work habits, some users will see it and others wont. We're actually taking a look at hiding that for the initial build in our issues at the moment. If you'd like to hold off on this PR until that's resolved, that's cool with me.

File: webpack.dev.js [hunk 1]
```diff
@@ -1,7 +1,18 @@
...
 
 module.exports = merge(common, {
   mode: 'development',
   devtool: 'inline-source-map',
+  entry: ['webpack-plugin-serve/client'],
+  plugins: [
+    new WebpackPluginServe({
+      host: 'localhost',
+      open: true,
+      static: join(__dirname, 'dist'),
```
Comment (smashercosmo) [inline webpack.dev.js]: Path can be taken from webpack config

**Comment by tr1s** [comment]
Merged, @shellscape @smashercosmo can you take a double look make sure it's all cool? 😴

Thanks again!

**Comment by shellscape** [comment]
I'll have a look at this tonight and let you know what I find

**Comment by smashercosmo** [comment]
@tr1s  Seems like your code doesn't accept hot updates anywhere. And in webpack-plugin-serve 'hot' option is true by default.

**Comment by shellscape** [comment]
@smashercosmo is correct. We can add `hmr: false` to the options to resolve that without having to modify the boilerplate.
